### PR TITLE
Add assert_type in dataclass transformer

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
[slack thread](https://flyte-org.slack.com/archives/CREL4QVAQ/p1661266231263149)

Add `assert_type()`in dataclass transformer, so we don't need to specify `type_hints` when using flyte remote.

We got an error when using dataclass as input in the flyte remote without passing `type_hints`, because `guess_python_type` is not equal to the user-defined dataclass type

```
@dataclass_json
@dataclass
class Foo(object):
    a: int = 0

@task
def t1(a: Foo):
    ...
```
In this example, we use guess_python_type (FooSchema) as our `expected_type` by default if using flyte remote to execute the above task.
However, FooSchema is created by flytekit and it's not equal to the user-defined dataclass (Foo), so [assert_type](https://github.com/flyteorg/flytekit/blob/aff19cb8bd59c5376cd638cef4ee39f05786a3a8/flytekit/core/type_engine.py#L95) raises an error.

We can override `assert_type` in the dataclass transformer, and check that the type of each attribute is the same as the `guess_type` by iterating all attributes in dataclass.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

The code to reproduce

```python
from dataclasses import dataclass
from typing import Optional
from dataclasses_json import dataclass_json
from flytekit import task, workflow, dynamic


@dataclass_json
@dataclass
class Args(object):
    max_iter: int = 0
    init_eval: float = 0
    max_eval: str = 0


@dataclass_json
@dataclass
class Schema(object):
    kwargs: Optional[Args] = None


@task
def t11(schema: Schema):
    print(schema)


@workflow
def wf3(schema: Schema = Schema(Args(1, 2.0, "hello"))):
    t11(schema=schema)


if __name__ == "__main__":
    wf3(schema=Schema(Args(1, 2.0, "hello")))
```

```python
from flytekit.remote import FlyteRemote
from flytekit.configuration import Config
from example_dataclass import Args, Schema

remote = FlyteRemote(
    config=Config.auto("/Users/kevin/.flyte/config-remote.yaml"),
    default_project="flytesnacks",
    default_domain="development",
)

flyte_workflow = remote.fetch_workflow(
    name="example_dataclass.wf3", project="flytesnacks", domain="development"
)

execution = remote.execute(
    flyte_workflow, inputs={"schema": Schema(Args(1, 2, "hello"))}, wait=False,
)
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2844

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -270,6 +270,46 @@ class Test():
...
+            if UnionTransformer.is_optional_type(original_type):
+                original_type = UnionTransformer.get_sub_type_in_optional(original_type)
+            if UnionTransformer.is_optional_type(expected_type):
+                expected_type = UnionTransformer.get_sub_type_in_optional(expected_type)
+
+            val = v.__getattribute__(f.name)
+            if dataclasses.is_dataclass(val):
+                self.assert_type(expected_type, val)
+            elif original_type != expected_type:
+                raise TypeTransformerFailedError(f"Type of Val '{original_type}' is not an instance of {expected_type}")
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:311]: is it easy to add a test for this line?

**Review by wild-endeavor — APPROVED** [review]
is there really nothing in dataclass json or json schema that will do this? we have to write our own code?

File: flytekit/core/type_engine.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/core/type_engine.py:311]: Updated it. 
I've tried to find a kind of function that can compare the attribute in dataclass, but I didn't find that.

**Review by wild-endeavor — APPROVED** [review]
